emerging_optimizers/psgd/psgd_utils.py

```diff
@@ -0,0 +1,168 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List
+
+import torch
+
+
+__all__ = [
+    "balance_q_in_place",
+    "norm_lower_bound_spd",
+    "norm_lower_bound_skew",
+]
+
+
+@torch.compile  # type: ignore[misc]
+def balance_q_in_place(Q_list: List[torch.Tensor]) -> None:
+    """Balance the dynamic ranges of kronecker factors in place to prevent numerical underflow or overflow.
+
+    Each tensor in `Q_list` is rescaled so that its maximum absolute entry
+    becomes the geometric mean of all factors original maxima. This preserves
+    the overall product of norms (and thus the scale of the Kronecker product)
+    while avoiding numerical underflow or overflow when factors have widely
+    differing magnitudes.
+
+    Given tensors :math:`Q_1, Q_2, \\ldots, Q_n`:
+
+    1. Compute max-absolute norms: :math:`\\|Q_i\\|_\\infty = \\max(|Q_i|)` for :math:`i = 1, \\ldots, n`
+    2. Compute geometric mean: :math:`g = \\left(\\prod_{i=1}^{n} \\|Q_i\\|_\\infty \\right)^{1/n}`
+    3. Rescale each tensor: :math:`Q_i \\leftarrow Q_i \\cdot \\frac{g}{\\|Q_i\\|_\\infty}`
+
+    This ensures :math:`\\|Q_i\\|_\\infty = g` for all :math:`i`, while preserving the norm of
+    the Kronecker product :math:`Q_1 \\otimes Q_2 \\otimes \\cdots \\otimes Q_n`.
+
+    Args:
+        Q_list: List of Q (e.g. the Kronecker factors), each tensor will be modified in place.
+
+    Returns:
+        None
+
+    """
+    if (order := len(Q_list)) <= 1:
+        return
```

Comment:
This should be a TypeError. Launching a function with empty list is generally considered wrong usage.

Also use `if list_name` to determine whether the list is empty.